### ISSUE TITLE
Can't get pouchDB to use SQLlite plugin in Cordova App.

### ISSUE THREAD

**Issue by irvins** [pouchdb-community/pouchdb-adapter-cordova-sqlite#22]
I keep seeing this error in my pouchdb-hellow-world barebones installation when testing in the phonegap developer app on my ios 9.3 , iphone 6s: 

 {"status":500,"name":"web_sql_went_bad","message":"unknown","reason":"SecurityError: DOM Exception 18"}

this is all i have in my onDeviceReady function right now

```

onDeviceReady: function() {
        var db      = new PouchDB("pouch.db");                         
        var websql  = new PouchDB("websqlpouch.db",{adapter: "websql"}); 
        var sqllite = new PouchDB("sqllitepouch.db",{adapter: "cordova-sqlite"); 
 
        if(db.adapter){
            $("#anypouch").text("PouchDB is working! Using adapter = " + db.adapter);
        }else{
            $("#anypouch").text("PouchDB nope!");
        }

        if(websql.adapter){
            $("#websql").text("webSQL is working! Using adapter = " + websql.adapter);
        }else{
            $("#websql").text("webSQL nope!")
        }

        sqllite.post({}).then(function (res) {
          return sqllite.get(res.id);
        }).then(function (doc) {
          alert('stored a document! ' + JSON.stringify(doc));
          alert('adapter is: ' + sqllite.adapter);
        }).catch(
            console.log.bind(console)
        );

        // if(sqllite.adapter){
        //     $("#sqllite").text("&#10003; SQLite plugin is working.");
        //     console.log("SQLitE Plugin OK" + sqllite.adapter);
        // }else{
        //     console.log("SQLitE Plugin nope")
        // }
    }
```

this is all i have in my index.html:
```

<body>
        <blockquote>
            <h3>He Needs His Mama</h3>
            <div id="anypouch">&#10007; PouchDB is not working.</div>
            <div id="websql">&#10007;  WebSQL is not supported.</div>
            <div id="sqllite">&#10007; SQLLite plugin is not supported.</div>
        </blockquote>
        <script type="text/javascript" src="cordova.js"></script>
        <script type="text/javascript" src="https://code.jquery.com/jquery-3.1.1.min.js"></script>             
        <script type="text/javascript" src="js/pouchdb-6.1.1.min.js"></script>
        <script type="text/javascript" src="js/pouchdb.cordova-sqlite.js"></script>

        <script type="text/javascript" src="js/index.js"></script>

    </body>

```

and here is what i have in my cordova plugins:
```

cordova plugins
cordova-plugin-sqlite-2 1.0.4 "SQLitePlugin"
cordova-plugin-websql 0.0.10 "Web SQL plugin"
cordova-plugin-whitelist 1.3.1 "Whitelist"
cordova-sqlite-storage 1.5.2 "Cordova sqlite storage plugin"


```

ive been googling for days and have hit a wall.  Any help would be appreciated.

i keep seeing the error with my actual app, which is why tried this barebones install to see if i can isolate it.  I only have the sql lite plugins + the sqllite adapter code.  it make me wonder if i ever had the sqllite plugin working at all.

also no matter how many plugins i install via the CLI

i still see this 

```
 PouchDB error: you must install a SQLite plugin in order for PouchDB to work on this platform. Options:
 - https://github.com/nolanlawson/cordova-plugin-sqlite-2
 - https://github.com/litehelpers/Cordova-sqlite-storage
 - https://github.com/Microsoft/cordova-plugin-websql
```

**Comment by nolanlawson** [comment]
If you are getting DOMException 18 it's typically because you're running out of space in the WebSQL adapter. According to the error you do not have the SQLite Plugin installed, i.e. at the moment you call `new PouchDB()` it's not able to find `cordova.sqlitePlugin`.

My guess is that your plugins are conflicting with each other, since you have `cordova-plugin-sqlite-2`, `cordova-plugin-websql`, and `cordova-sqlite-storage` all installed at the same time.

Can you check in the debugger and see if `cordova.sqlitePlugin` is available when you call `new PouchDB()`?

**Comment by Gengjun-VVU** [comment]
Hi,
I have the same issue using Ionic 2 (v2.2.0), I followed the tutorial on [Ashteya Biharisingh's blog](https://gonehybrid.com/how-to-use-pouchdb-sqlite-for-local-storage-in-ionic-2/), for the provider, I have the following code:

    import { Injectable } from '@angular/core';
    import { Http } from '@angular/http';
    import 'rxjs/add/operator/map';
    import PouchDB from 'pouchdb';

    @Injectable()
    export class BirthdayService {

    private _db;
    private _birthdays;

    constructor(public http: Http) {
      console.log('Hello BirthdayService Provider');
    }
  
    initDB() {
      var PouchAdapterCordovaSqlite = require('pouchdb-adapter-cordova-sqlite');
      PouchDB.plugin(PouchAdapterCordovaSqlite);
      this._db = new PouchDB('birthday', {adapter: 'cordova-sqlite'});
      this._db.info().then(console.log.bind(console));
    }
  
    add(birthday) {
      return this._db.post(birthday);
    }
  
    update(birthday) {
      return this._db.put(birthday);
    }
  
    delete(birthday) {  
      return this._db.remove(birthday);
    }
  
    getAll() {  
    
      if (!this._birthdays) {
        return this._db.allDocs({ include_docs: true})
            .then(docs => {

                  // Each row has a .doc object and we just want to send an 
                  // array of birthday objects back to the calling controller,
                  // so let's map the array to contain just the .doc objects.

                  this._birthdays = docs.rows.map(row => {
                      // Dates are not automatically converted from a string.
                      row.doc.Date = new Date(row.doc.Date);
                      return row.doc;
                  });

                  // Listen for changes on the database.
                  this._db.changes({ live: true, since: 'now', include_docs: true})
                               .on('change', this.onDatabaseChange);

                  return this._birthdays;
              });
      } else {
          // Return cached data as a promise
          return Promise.resolve(this._birthdays);
      }
    }
  
    private onDatabaseChange = (change) => {  
    var index = this.findIndex(this._birthdays, change.id);
    var birthday = this._birthdays[index];

    if (change.deleted) {
        if (birthday) {
            this._birthdays.splice(index, 1); // delete
        }
    } else {
        change.doc.Date = new Date(change.doc.Date);
        if (birthday && birthday._id === change.id) {
            this._birthdays[index] = change.doc; // update
        } else {
            this._birthdays.splice(index, 0, change.doc) // insert
        }
      }
    }

      // Binary search, the array is by default sorted by _id.
      private findIndex(array, id) {  
        var low = 0, high = array.length, mid;
        while (low < high) {
          mid = (low + high) >>> 1;
          array[mid]._id < id ? low = mid + 1 : high = mid
        }
        return low;
      }
    }

The reason why I can use require() is because I used the following command to add types:
    `$ npm install @types/node --save`

When I start `$ ionic serve`, I got the same error:

> PouchDB error: you must install a SQLite plugin in order for PouchDB to work on this platform. Options:
> - https://github.com/nolanlawson/cordova-plugin-sqlite-2
> - https://github.com/litehelpers/Cordova-sqlite-storage
> - https://github.com/Microsoft/cordova-plugin-websql

Even though I have everything listed above installed via ionic CLI or npm CLI.

Since the error happens here, the PouchDB falls back to use WebSQL. I was wondering if I need to build the app with specific platform, say, android or ios, then the error may disappear?

I have tried the app on Ionic View app, it won't save the data. I really appreciate if you can provide an example using sqlite with Ionic 2. Thanks.

GJ Wu

**Comment by nolanlawson** [comment]
The issue may be that the SQLite plugin is not loaded when PouchDB is initialized. PouchDB checks for window.sqlitePlugin at the *moment* you call new PouchDB(). You may need to wait for the plugin to be initialized by waiting for the Cordova “deviceready” event.

**Comment by chaje** [comment]
@Gengjun-VVU can you try this:
 this._db = new PouchDB('birthday.');
Turns out that when forcing the websql adapter the db instance is not visible in PouchDB Inspector; using the default db adapter makes the db instance appear well.

**Comment by irvins** [comment]
Sorry Nolan, never followed up on this.  But it seemed to go away for a bit, but now it rears its ugly head again.

i only have one sqlite-2 plugin and i still get this error

PouchDB error: you must install a SQLite plugin in order for PouchDB to work on this platform. Options:
https://github.com/nolanlawson/cordova-plugin-sqlite-2
https://github.com/litehelpers/Cordova-sqlite-storage
https://github.com/Microsoft/cordova-plugin-websql 

ive emptied local storage and emptied my local pouchdb.   

on the first line of my onDeviceReady: function(){
//i put
var newdb = new PouchDB();

console.log("did it work?");
console.log(newdb);

}

never get to the next line.

sorry.

**Comment by mnewmedia** [comment]
### Prepare
1. You need [PouchDB](https://github.com/pouchdb/pouchdb)
2. You need [cordova-plugin-sqlite-2](https://github.com/nolanlawson/cordova-plugin-sqlite-2)
3. You need [pouchdb-adapter-cordova-sqlite](https://github.com/pouchdb-community/pouchdb-adapter-cordova-sqlite)

### Usage
```js
// load cordova adapter
var PouchAdapterCordovaSqlite = require('pouchdb-adapter-cordova-sqlite');
// register cordova adapter
PouchDB.plugin(PouchAdapterCordovaSqlite);

// register event to be notified when cordova and so cordova-sqlite-plugin-2 is ready
document.addEventListener("deviceready", onDeviceReady, false);

// when cordova is ready
function onDeviceReady() {
    // then create the database
    var db = new PouchDB('myDB', {adapter: 'cordova-sqlite'});
}
```

**Comment by irvins** [comment]
THanks mnewmedia.  That is how i set up my app.  i followed that page exactly.  and it seemed to be working fine for months.   

now just trying to do var db = new PouchDB('myDB', {adapter: 'cordova-sqlite'});
gets me this error.


PouchDB error: you must install a SQLite plugin in order for PouchDB to work on this platform. Options:
 - https://github.com/nolanlawson/cordova-plugin-sqlite-2
 - https://github.com/litehelpers/Cordova-sqlite-storage
 - https://github.com/Microsoft/cordova-plugin-websql
WebSQL threw an error QuotaExceededError (DOM Exception 22): The quota has been exceeded.
{"status":500,"name":"web_sql_went_bad","message":"unknown","reason":"The quota has been exceeded."}

**Comment by mnewmedia** [comment]
For me it's working like i showed above.. if not for you, you have to provide a github repository where we can reproduce the error. Else I can not help you..

**Comment by humblecoder** [comment]
@mnewmedia where are you declaring the "PouchDB" object?  I'm currently using Vue with OnsenUI.  I've attempted to trigger the loading in "Vue.cordova.on" as well as "$ons.ready" and I still receive:  "PouchDB error: you must install a SQLite plugin in order for PouchDB to work on this platform."

FYI, I've tried using all 3 plugins to no avail.  For what it's worth, I'm using:

- Cordova 7.0.1
- Monaca 2.2.2
- VueJS
- Vuex
- PouchDB 6.2.0
- OnsenUI

**Comment by mnewmedia** [comment]
@humblecoder I don't know about `Vue.cordova.on` or `$ons.ready` but I would go for the original `deviceready` from cordova no matter which framework you are using.

**Comment by humblecoder** [comment]
@mnewmedia it appears that vanilla Cordova simply relies on "cordova.js" (which Monaca already imports), then using "document.addEventListener('deviceready',...)" which doesn't seem to fire consistently -- Vue.cordova.on an d$ons.ready seem to fire consistently and everything else is in order.  Even vanilla PouchDB works fine.  The only issue I'm having is the actual loading of the PouchDB plugins.

Would you mind ELI5 (#cringe) and showing every relevant step in your setup?  For example:

1. npm install [plugin]
2. inside "main.js" declare "var PouchDB = require('pouchdb')"
3.  ...
4.  ...

**Comment by marius-avram** [comment]
I confirm this issue to be present in ionic 2. Probably it works just fine in ionic 1 or cordova.
I'm not very familiar with it ionic 2. 

But I also tried initialising PouchDB after Platform.ready() using the Platform service but I had no success in making it work. I still receive the same error:

PouchDB error: you must install a SQLite plugin in order for PouchDB to work on this platform. Options:
 - https://github.com/nolanlawson/cordova-plugin-sqlite-2
 - https://github.com/litehelpers/Cordova-sqlite-storage
 - https://github.com/Microsoft/cordova-plugin-websql

Such a deal breaker.

**Comment by zhaoxin** [comment]
Same issue on ionic2. The code got stuck on calling "new PouchDB('my.db', {adapter: 'cordova-sqlite'});",
while removing {adapter: 'cordova-sqlite'} make it run through.

**Comment by zhaoxin** [comment]
Downgrade to pouchdb 6.2.0 and issue solved. PouchDB related lines in my package.json, FYI:
"pouchdb": "6.2.0",
"pouchdb-adapter-utils": "6.2.0",
"pouchdb-json": "6.2.0",
"pouchdb-adapter-cordova-sqlite": "2.0.2",
"pouchdb-adapter-websql-core": "6.2.0",
"pouchdb-browser": "6.2.0",
"pouchdb-core": "6.2.0",

**Comment by davidanton1d** [comment]
Hi, I had the same problem and found a possible answer/workaround in [**this ionic forum thread**](https://forum.ionicframework.com/t/update-to-3-5-3-and-pouchdb/98724/26):

Basically, downgrade to 6.2.0 and adding the package `"pouchdb-adapter-websql-core": "6.2.0"` made it work for me. (I also switched to the cordova-plugin-sqlite-2, don't know if that had did any difference?)

Could there be a package missing in the pouchdb-adapter-cordova-sqlite dependency list?

I'm getting past the "you must install a SQLite plugin" error, but instead pouchdb.sync() doesn't work. Anyone tried sync() with cordova-sqlite?

**Comment by paolosanchi** [comment]
This is stale, hope it has been solved in the latest releases
